File: go/task/19tree.go
package task

import (
	"fmt"

	"github.com/pkg/errors"
	"github.com/shahzodshafizod/ptaskbook/go/pkg/checker"
	"github.com/shahzodshafizod/ptaskbook/go/pkg/scanner"
	"github.com/shahzodshafizod/ptaskbook/go/pkg/tree"
)

type treeTask struct {
	dataPath string
	name     string
	count    int
	scanner  scanner.Scanner
	checker  checker.Checker
}

func NewTreeTask(pathPrefix string) Task {
	return &treeTask{
		dataPath: pathPrefix + "/19tree/Tree%03d/%03d/%03d",
		name:     "Tree",
		count:    100,
	}
}

func (tt *treeTask) GetCount() int { return tt.count }

func (tt *treeTask) GetName() string { return tt.name }

func (tt *treeTask) SetData(taskNo int, testNo int) error {
	var filePath = fmt.Sprintf(tt.dataPath, taskNo, testNo, testNo)
	scanner, err := scanner.New(filePath + ".dat")
	if err != nil {
		return errors.Wrap(err, "scanner.New")
	}
	checker, err := checker.New(filePath + ".ans")
	if err != nil {
		return errors.Wrap(err, "checker.New")
	}
	tt.scanner = scanner
	tt.checker = checker
	return nil
}

func (tt *treeTask) ScannerEOF() bool { return tt.scanner.EOF() }

func (tt *treeTask) CheckerEOF() bool { return tt.checker.EOF() }

func (tt *treeTask) CleanData() {
	tt.scanner.RemoveFiles()
	tt.checker.RemoveFiles()
}

func (tt *treeTask) GetFn(taskNo int) func() bool {
	switch taskNo {
	case 1:
		return tt.tree1
	case 2:
		return tt.tree2
	case 3:
		return tt.tree3
	case 4:
		return tt.tree4
	case 5:
		return tt.tree5
	case 6:
		return tt.tree6
	case 7:
		return tt.tree7
	case 8:
		return tt.tree8
	case 9:
		return tt.tree9
	case 10:
		return tt.tree10
	case 11:
		return tt.tree11
	case 12:
		return tt.tree12
	case 13:
		return tt.tree13
	case 14:
		return tt.tree14
	case 15:
		return tt.tree15
	case 16:
		return tt.tree16
	case 17:
		return tt.tree17
	case 18:
		return tt.tree18
	case 19:
		return tt.tree19
	case 20:
		return tt.tree20
	case 21:
		return tt.tree21
	case 22:
		return tt.tree22
	case 23:
		return tt.tree23
	case 24:
		return tt.tree24
	case 25:
		return tt.tree25
	case 26:
		return tt.tree26
	case 27:
		return tt.tree27
	case 28:
		return tt.tree28
	case 29:
		return tt.tree29
	case 30:
		return tt.tree30
	case 31:
		return tt.tree31
	case 32:
		return tt.tree32
	case 33:
		return tt.tree33
	case 34:
		return tt.tree34
	case 35:
		return tt.tree35
	case 36:
		return tt.tree36
	case 37:
		return tt.tree37
	case 38:
		return tt.tree38
	case 39:
		return tt.tree39
	case 40:
		return tt.tree40
	case 41:
		return tt.tree41
	case 42:
		return tt.tree42
	case 43:
		return tt.tree43
	case 44:
		return tt.tree44
	case 45:
		return tt.tree45
	case 46:
		return tt.tree46
	case 47:
		return tt.tree47
	case 48:
		return tt.tree48
	case 49:
		return tt.tree49
	case 50:
		return tt.tree50
	case 51:
		return tt.tree51
	case 52:
		return tt.tree52
	case 53:
		return tt.tree53
	case 54:
		return tt.tree54
	case 55:
		return tt.tree55
	case 56:
		return tt.tree56
	case 57:
		return tt.tree57
	case 58:
		return tt.tree58
	case 59:
		return tt.tree59
	case 60:
		return tt.tree60
	case 61:
		return tt.tree61
	case 62:
		return tt.tree62
	case 63:
		return tt.tree63
	case 64:
		return tt.tree64
	case 65:
		return tt.tree65
	case 66:
		return tt.tree66
	case 67:
		return tt.tree67
	case 68:
		return tt.tree68
	case 69:
		return tt.tree69
	case 70:
		return tt.tree70
	case 71:
		return tt.tree71
	case 72:
		return tt.tree72
	case 73:
		return tt.tree73
	case 74:
		return tt.tree74
	case 75:
		return tt.tree75
	case 76:
		return tt.tree76
	case 77:
		return tt.tree77
	case 78:
		return tt.tree78
	case 79:
		return tt.tree79
	case 80:
		return tt.tree80
	case 81:
		return tt.tree81
	case 82:
		return tt.tree82
	case 83:
		return tt.tree83
	case 84:
		return tt.tree84
	case 85:
		return tt.tree85
	case 86:
		return tt.tree86
	case 87:
		return tt.tree87
	case 88:
		return tt.tree88
	case 89:
		return tt.tree89
	case 90:
		return tt.tree90
	case 91:
		return tt.tree91
	case 92:
		return tt.tree92
	case 93:
		return tt.tree93
	case 94:
		return tt.tree94
	case 95:
		return tt.tree95
	case 96:
		return tt.tree96
	case 97:
		return tt.tree97
	case 98:
		return tt.tree98
	case 99:
		return tt.tree99
	case 100:
		return tt.tree100
	default:
		return nil
	}
}

func (tt *treeTask) tree1() bool {
	var p1 = tt.scanner.NextTree()
	if !tt.checker.CompareInt(p1.Data, p1.Left.Data, p1.Right.Data) {
		return false
	}
	return tt.checker.CompareTree(p1, p1.Left, p1.Right)
}

func (tt *treeTask) tree2() bool {
	var p1 = tt.scanner.NextTree()
	var nodesCount = tree.GetCount(p1, 0, 0, "all")
	return tt.checker.CompareInt(nodesCount)
}

func (tt *treeTask) tree3() bool {
	var p1 = tt.scanner.NextTree()
	var k = tt.scanner.NextInt()
	var kCounts = tree.GetCount(p1, k, 0, "data")
	return tt.checker.CompareInt(kCounts)
}

func (tt *treeTask) tree4() bool {
	var p1 = tt.scanner.NextTree()
	var sum = tree.GetSum(p1, 0, 0, "all")
	return tt.checker.CompareInt(sum)
}

func (tt *treeTask) tree5() bool {
	var p1 = tt.scanner.NextTree()
	var leftNodesCount = tree.GetCount(p1, 0, 0, "left")
	return tt.checker.CompareInt(leftNodesCount)
}

func (tt *treeTask) tree6() bool {
	var p1 = tt.scanner.NextTree()
	var leavesCount = tree.GetCount(p1, 0, 0, "leaf")
	return tt.checker.CompareInt(leavesCount)
}

func (tt *treeTask) tree7() bool {
	var p1 = tt.scanner.NextTree()
	var leavesSum = tree.GetSum(p1, 0, 0, "leaf")
	return tt.checker.CompareInt(leavesSum)
}

func (tt *treeTask) tree8() bool {
	var p1 = tt.scanner.NextTree()
	var rightLeavesCount = tree.GetCount(p1, 0, 0, "right", "leaf")
	return tt.checker.CompareInt(rightLeavesCount)
}

func (tt *treeTask) tree9() bool {
	return false
}

func (tt *treeTask) tree10() bool {
	return false
}

func (tt *treeTask) tree11() bool {
	return false
}

func (tt *treeTask) tree12() bool {
	return false
}

func (tt *treeTask) tree13() bool {
	return false
}

func (tt *treeTask) tree14() bool {
	return false
}

func (tt *treeTask) tree15() bool {
	return false
}

func (tt *treeTask) tree16() bool {
	return false
}

func (tt *treeTask) tree17() bool {
	return false
}

func (tt *treeTask) tree18() bool {
	return false
}

func (tt *treeTask) tree19() bool {
	return false
}

func (tt *treeTask) tree20() bool {
	return false
}

func (tt *treeTask) tree21() bool {
	return false
}

func (tt *treeTask) tree22() bool {
	return false
}

func (tt *treeTask) tree23() bool {
	return false
}

func (tt *treeTask) tree24() bool {
	return false
}

func (tt *treeTask) tree25() bool {
	return false
}

func (tt *treeTask) tree26() bool {
	return false
}

func (tt *treeTask) tree27() bool {
	return false
}

func (tt *treeTask) tree28() bool {
	return false
}

func (tt *treeTask) tree29() bool {
	return false
}

func (tt *treeTask) tree30() bool {
	return false
}

func (tt *treeTask) tree31() bool {
	return false
}

func (tt *treeTask) tree32() bool {
	return false
}

func (tt *treeTask) tree33() bool {
	return false
}

func (tt *treeTask) tree34() bool {
	return false
}

func (tt *treeTask) tree35() bool {
	return false
}

func (tt *treeTask) tree36() bool {
	return false
}

func (tt *treeTask) tree37() bool {
	return false
}

func (tt *treeTask) tree38() bool {
	return false
}

func (tt *treeTask) tree39() bool {
	return false
}

func (tt *treeTask) tree40() bool {
	return false
}

func (tt *treeTask) tree41() bool {
	return false
}

func (tt *treeTask) tree42() bool {
	return false
}

func (tt *treeTask) tree43() bool {
	return false
}

func (tt *treeTask) tree44() bool {
	return false
}

func (tt *treeTask) tree45() bool {
	return false
}

func (tt *treeTask) tree46() bool {
	return false
}

func (tt *treeTask) tree47() bool {
	return false
}

func (tt *treeTask) tree48() bool {
	return false
}

func (tt *treeTask) tree49() bool {
	return false
}

func (tt *treeTask) tree50() bool {
	return false
}

func (tt *treeTask) tree51() bool {
	return false
}

func (tt *treeTask) tree52() bool {
	return false
}

func (tt *treeTask) tree53() bool {
	return false
}

func (tt *treeTask) tree54() bool {
	return false
}

func (tt *treeTask) tree55() bool {
	return false
}

func (tt *treeTask) tree56() bool {
	return false
}

func (tt *treeTask) tree57() bool {
	return false
}

func (tt *treeTask) tree58() bool {
	return false
}

func (tt *treeTask) tree59() bool {
	return false
}

func (tt *treeTask) tree60() bool {
	return false
}

func (tt *treeTask) tree61() bool {
	return false
}

func (tt *treeTask) tree62() bool {
	return false
}

func (tt *treeTask) tree63() bool {
	return false
}

func (tt *treeTask) tree64() bool {
	return false
}

func (tt *treeTask) tree65() bool {
	return false
}

func (tt *treeTask) tree66() bool {
	return false
}

func (tt *treeTask) tree67() bool {
	return false
}

func (tt *treeTask) tree68() bool {
	return false
}

func (tt *treeTask) tree69() bool {
	return false
}

func (tt *treeTask) tree70() bool {
	return false
}

func (tt *treeTask) tree71() bool {
	return false
}

func (tt *treeTask) tree72() bool {
	return false
}

func (tt *treeTask) tree73() bool {
	return false
}

func (tt *treeTask) tree74() bool {
	return false
}

func (tt *treeTask) tree75() bool {
	return false
}

func (tt *treeTask) tree76() bool {
	return false
}

func (tt *treeTask) tree77() bool {
	return false
}

func (tt *treeTask) tree78() bool {
	return false
}

func (tt *treeTask) tree79() bool {
	return false
}

func (tt *treeTask) tree80() bool {
	return false
}

func (tt *treeTask) tree81() bool {
	return false
}

func (tt *treeTask) tree82() bool {
	return false
}

func (tt *treeTask) tree83() bool {
	return false
}

func (tt *treeTask) tree84() bool {
	return false
}

func (tt *treeTask) tree85() bool {
	return false
}

func (tt *treeTask) tree86() bool {
	return false
}

func (tt *treeTask) tree87() bool {
	return false
}

func (tt *treeTask) tree88() bool {
	return false
}

func (tt *treeTask) tree89() bool {
	return false
}

func (tt *treeTask) tree90() bool {
	return false
}

func (tt *treeTask) tree91() bool {
	return false
}

func (tt *treeTask) tree92() bool {
	return false
}

func (tt *treeTask) tree93() bool {
	return false
}

func (tt *treeTask) tree94() bool {
	return false
}

func (tt *treeTask) tree95() bool {
	return false
}

func (tt *treeTask) tree96() bool {
	return false
}

func (tt *treeTask) tree97() bool {
	return false
}

func (tt *treeTask) tree98() bool {
	return false
}

func (tt *treeTask) tree99() bool {
	return false
}

func (tt *treeTask) tree100() bool {
	return false
}

File: kt/task/19tree.kt
package task

class TreeTask {

    var datScanner = scanner.Scanner()
    var ansScanner = scanner.Scanner()

    fun solve(pathPrefix: String) {
        for (taskNo in 1..100) {
            var taskNoStr = "%03d".format(taskNo)
            var fileName: String
            for (testNo in 1..100) {
                var testNoStr = "%03d".format(testNo)
                fileName = "$pathPrefix/19tree/Tree$taskNoStr/$testNoStr/$testNoStr"
                if (!datScanner.changeReader("$fileName.dat")) {
                    println("Tree$taskNoStr couldn't open datScanner source")
                    return
                }
                if (!ansScanner.changeReader("$fileName.ans")) {
                    println("Tree$taskNoStr couldn't open ansScanner source")
                    return
                }
                val ok = when (taskNo) {
                    1 -> this.tree1()
                    2 -> this.tree2()
                    3 -> this.tree3()
                    4 -> this.tree4()
                    5 -> this.tree5()
                    6 -> this.tree6()
                    7 -> this.tree7()
                    8 -> this.tree8()
                    9 -> this.tree9()
                    10 -> this.tree10()
                    11 -> this.tree11()
                    12 -> this.tree12()
                    13 -> this.tree13()
                    14 -> this.tree14()
                    15 -> this.tree15()
                    16 -> this.tree16()
                    17 -> this.tree17()
                    18 -> this.tree18()
                    19 -> this.tree19()
                    20 -> this.tree20()
                    21 -> this.tree21()
                    22 -> this.tree22()
                    23 -> this.tree23()
                    24 -> this.tree24()
                    25 -> this.tree25()
                    26 -> this.tree26()
                    27 -> this.tree27()
                    28 -> this.tree28()
                    29 -> this.tree29()
                    30 -> this.tree30()
                    31 -> this.tree31()
                    32 -> this.tree32()
                    33 -> this.tree33()
                    34 -> this.tree34()
                    35 -> this.tree35()
                    36 -> this.tree36()
                    37 -> this.tree37()
                    38 -> this.tree38()
                    39 -> this.tree39()
                    40 -> this.tree40()
                    41 -> this.tree41()
                    42 -> this.tree42()
                    43 -> this.tree43()
                    44 -> this.tree44()
                    45 -> this.tree45()
                    46 -> this.tree46()
                    47 -> this.tree47()
                    48 -> this.tree48()
                    49 -> this.tree49()
                    50 -> this.tree50()
                    51 -> this.tree51()
                    52 -> this.tree52()
                    53 -> this.tree53()
                    54 -> this.tree54()
                    55 -> this.tree55()
                    56 -> this.tree56()
                    57 -> this.tree57()
                    58 -> this.tree58()
                    59 -> this.tree59()
                    60 -> this.tree60()
                    61 -> this.tree61()
                    62 -> this.tree62()
                    63 -> this.tree63()
                    64 -> this.tree64()
                    65 -> this.tree65()
                    66 -> this.tree66()
                    67 -> this.tree67()
                    68 -> this.tree68()
                    69 -> this.tree69()
                    70 -> this.tree70()
                    71 -> this.tree71()
                    72 -> this.tree72()
                    73 -> this.tree73()
                    74 -> this.tree74()
                    75 -> this.tree75()
                    76 -> this.tree76()
                    77 -> this.tree77()
                    78 -> this.tree78()
                    79 -> this.tree79()
                    80 -> this.tree80()
                    81 -> this.tree81()
                    82 -> this.tree82()
                    83 -> this.tree83()
                    84 -> this.tree84()
                    85 -> this.tree85()
                    86 -> this.tree86()
                    87 -> this.tree87()
                    88 -> this.tree88()
                    89 -> this.tree89()
                    90 -> this.tree90()
                    91 -> this.tree91()
                    92 -> this.tree92()
                    93 -> this.tree93()
                    94 -> this.tree94()
                    95 -> this.tree95()
                    96 -> this.tree96()
                    97 -> this.tree97()
                    98 -> this.tree98()
                    99 -> this.tree99()
                    100 -> this.tree100()
                    else -> false
                }
                scanner.removeFiles()
                if (!ok) {
                    println("Tree$taskNoStr Test#$testNoStr has not been tested")
                    return
                }
                if (!datScanner.eof()) {
                    println("Tree$taskNoStr Test#$testNoStr input is not enough")
                    return
                }
                if (!ansScanner.eof()) {
                    println("Tree$taskNoStr Test#$testNoStr output is not enough")
                    return
                }
            }
            println("Tree$taskNoStr has been tested!")
        }
    }

    private fun tree1(): Boolean {
        return false
    }

    private fun tree2(): Boolean {
        return false
    }

    private fun tree3(): Boolean {
        return false
    }

    private fun tree4(): Boolean {
        return false
    }

    private fun tree5(): Boolean {
        return false
    }

    private fun tree6(): Boolean {
        return false
    }

    private fun tree7(): Boolean {
        return false
    }

    private fun tree8(): Boolean {
        return false
    }

    private fun tree9(): Boolean {
        return false
    }

    private fun tree10(): Boolean {
        return false
    }

    private fun tree11(): Boolean {
        return false
    }

    private fun tree12(): Boolean {
        return false
    }

    private fun tree13(): Boolean {
        return false
    }

    private fun tree14(): Boolean {
        return false
    }

    private fun tree15(): Boolean {
        return false
    }

    private fun tree16(): Boolean {
        return false
    }

    private fun tree17(): Boolean {
        return false
    }

    private fun tree18(): Boolean {
        return false
    }

    private fun tree19(): Boolean {
        return false
    }

    private fun tree20(): Boolean {
        return false
    }

    private fun tree21(): Boolean {
        return false
    }

    private fun tree22(): Boolean {
        return false
    }

    private fun tree23(): Boolean {
        return false
    }

    private fun tree24(): Boolean {
        return false
    }

    private fun tree25(): Boolean {
        return false
    }

    private fun tree26(): Boolean {
        return false
    }

    private fun tree27(): Boolean {
        return false
    }

    private fun tree28(): Boolean {
        return false
    }

    private fun tree29(): Boolean {
        return false
    }

    private fun tree30(): Boolean {
        return false
    }

    private fun tree31(): Boolean {
        return false
    }

    private fun tree32(): Boolean {
        return false
    }

    private fun tree33(): Boolean {
        return false
    }

    private fun tree34(): Boolean {
        return false
    }

    private fun tree35(): Boolean {
        return false
    }

    private fun tree36(): Boolean {
        return false
    }

    private fun tree37(): Boolean {
        return false
    }

    private fun tree38(): Boolean {
        return false
    }

    private fun tree39(): Boolean {
        return false
    }

    private fun tree40(): Boolean {
        return false
    }

    private fun tree41(): Boolean {
        return false
    }

    private fun tree42(): Boolean {
        return false
    }

    private fun tree43(): Boolean {
        return false
    }

    private fun tree44(): Boolean {
        return false
    }

    private fun tree45(): Boolean {
        return false
    }

    private fun tree46(): Boolean {
        return false
    }

    private fun tree47(): Boolean {
        return false
    }

    private fun tree48(): Boolean {
        return false
    }

    private fun tree49(): Boolean {
        return false
    }

    private fun tree50(): Boolean {
        return false
    }

    private fun tree51(): Boolean {
        return false
    }

    private fun tree52(): Boolean {
        return false
    }

    private fun tree53(): Boolean {
        return false
    }

    private fun tree54(): Boolean {
        return false
    }

    private fun tree55(): Boolean {
        return false
    }

    private fun tree56(): Boolean {
        return false
    }

    private fun tree57(): Boolean {
        return false
    }

    private fun tree58(): Boolean {
        return false
    }

    private fun tree59(): Boolean {
        return false
    }

    private fun tree60(): Boolean {
        return false
    }

    private fun tree61(): Boolean {
        return false
    }

    private fun tree62(): Boolean {
        return false
    }

    private fun tree63(): Boolean {
        return false
    }

    private fun tree64(): Boolean {
        return false
    }

    private fun tree65(): Boolean {
        return false
    }

    private fun tree66(): Boolean {
        return false
    }

    private fun tree67(): Boolean {
        return false
    }

    private fun tree68(): Boolean {
        return false
    }

    private fun tree69(): Boolean {
        return false
    }

    private fun tree70(): Boolean {
        return false
    }

    private fun tree71(): Boolean {
        return false
    }

    private fun tree72(): Boolean {
        return false
    }

    private fun tree73(): Boolean {
        return false
    }

    private fun tree74(): Boolean {
        return false
    }

    private fun tree75(): Boolean {
        return false
    }

    private fun tree76(): Boolean {
        return false
    }

    private fun tree77(): Boolean {
        return false
    }

    private fun tree78(): Boolean {
        return false
    }

    private fun tree79(): Boolean {
        return false
    }

    private fun tree80(): Boolean {
        return false
    }

    private fun tree81(): Boolean {
        return false
    }

    private fun tree82(): Boolean {
        return false
    }

    private fun tree83(): Boolean {
        return false
    }

    private fun tree84(): Boolean {
        return false
    }

    private fun tree85(): Boolean {
        return false
    }

    private fun tree86(): Boolean {
        return false
    }

    private fun tree87(): Boolean {
        return false
    }

    private fun tree88(): Boolean {
        return false
    }

    private fun tree89(): Boolean {
        return false
    }

    private fun tree90(): Boolean {
        return false
    }

    private fun tree91(): Boolean {
        return false
    }

    private fun tree92(): Boolean {
        return false
    }

    private fun tree93(): Boolean {
        return false
    }

    private fun tree94(): Boolean {
        return false
    }

    private fun tree95(): Boolean {
        return false
    }

    private fun tree96(): Boolean {
        return false
    }

    private fun tree97(): Boolean {
        return false
    }

    private fun tree98(): Boolean {
        return false
    }

    private fun tree99(): Boolean {
        return false
    }

    private fun tree100(): Boolean {
        return false
    }
}